tempesta_fw/ss_skb.c

```diff
@@ -533,80 +615,74 @@ __split_pgfrag_del(struct sk_buff *skb, int i, int off, int len, TfwStr *it)
 	 * Make room for a fragment right after the @i fragment
 	 * to move the tail part of data there.
 	 */
-	if (__extend_pgfrags(skb, i + 1, 1, it))
+	if (__extend_pgfrags(skb, bskb, i + 1, 1))
 		return -EFAULT;
 
 	/* Find the SKB for tail data. */
-	skb_dst = (i < MAX_SKB_FRAGS - 1) ? skb : it->skb;
+	skb_dst = (i < MAX_SKB_FRAGS - 1) ? skb : __next_skb_frag(skb, bskb);
 
 	/* Calculate the length of the tail part. */
 	tail_len = skb_frag_size(frag) - off - len;
 
-	/* Trim the fragment with the head part. */
-	skb_frag_size_sub(frag, len + tail_len);
-
 	/* Make the fragment with the tail part. */
 	i = (i + 1) % MAX_SKB_FRAGS;
 	__skb_fill_page_desc(skb_dst, i, skb_frag_page(frag),
 			     frag->page_offset + off + len, tail_len);
 	__skb_frag_ref(frag);
 
+	/* Trim the fragment with the head part. */
+	skb_frag_size_sub(frag, len + tail_len);
+
 	/* Adjust SKB data lengths. */
-	ss_skb_adjust_data_len(skb, -len);
 	if (skb != skb_dst) {
-		ss_skb_adjust_data_len(skb, -tail_len);
+		if (skb != bskb)
+			ss_skb_adjust_data_len(skb, -tail_len);
 		ss_skb_adjust_data_len(skb_dst, tail_len);
 	}
+	ss_skb_adjust_data_len(skb, -len);
+	ss_bskb_adjust_len(skb, bskb, -len);
 
-	/* Get the SKB and the address of data after the deleted area. */
-	it->flags = (skb != skb_dst);
+	/* Get the SKB and the address for data after the deleted data. */
 	it->ptr = skb_frag_address(&skb_shinfo(skb_dst)->frags[i]);
+	it->skb = bskb;
 	return 0;
 
 lookup_next_ptr:
-	/* Get the next fragment after the deleted fragment. */
-	if (i < si->nr_frags)
+	/* Get the address of data after the deleted data. */
+	if (i < si->nr_frags) {
 		it->ptr = skb_frag_address(&si->frags[i]);
+		it->skb = bskb;
+	} else {
+		struct sk_buff *dskb;
+		it->skb = __next_skb(skb, &bskb, &dskb);
+		it->ptr = __skb_data_address(dskb, bskb);
+	}
```

Comment:
copy&paste from `__split_linear_data()`. This should be done as a function

        static inline void
        __lookup_next_ptr(struct sk_buff *skb, struct sk_buff *bskb, TfwStr *it, int i)
        {
                struct skb_shared_info *si = skb_shinfo(skb);
                if (i < si->nr_frags) {                                                
                        it->ptr = skb_frag_address(&si->frags[i]);                     
                        it->skb = bskb;                                                
                } else {                                                               
                        struct sk_buff *dskb;                                          
                        it->skb = __next_skb(skb, &bskb, &dskb);                       
                        it->ptr = __skb_data_address(dskb, bskb);                      
                }               
        }

and should be called with `0` and `i` as last argument correspondingly.